### ISSUE TITLE
Gatway subscriptions doesn't execute subscribe on service side second time

### ISSUE THREAD

**Issue by 0x77dev** [mercurius-js/mercurius#424]
```asciichart
┌─────────────────────────────────────┐
│Mercurius Service                    │
│                                     │
│Subscription.testEvent               │
│                                     │
│() => subscibe(`testEvent-${userId}`)│
└────────▲────────────────────────────┘
         │
         │
         │
         │
         │
┌────────┴────────┐
│Mercurius Gateway│
└─▲──────────▲────┘
  │          │
  │          │
  │userId: 1 │userId: 2
  │          │
 ┌┴─────┐    │       ┌──────┐
 │Client│    └───────┤Client│
 └──────┘            └──────┘
```

Both clients subscribe to `Subscription.testEvent` through gateway, but on a service side `subscribe` executes only once, thus they both receive the same connection. 

As a result, `userId: 1` gets `userId: 2` events and `userId: 2` either does not receive any, or receives same events as `userId: 1`.

If clients subscribe to service directly without gateway it works as expected: `userId: 1` gets its events, the same as `userId: 2`.

**Comment by mcollina** [comment]
Can you write a small example to demonstrate your issue (both client and server, check the unit tests)? Would you like to send  a PR to fix it?

**Comment by 0x77dev** [comment]
### reproduce example

https://gist.github.com/0x77dev/662309d334348758e19af329d2d9b0ec

### without gateway

<img width="1416" alt="Screenshot 2021-03-16 at 15 46 19" src="https://user-images.githubusercontent.com/46429701/111319628-e4517400-866e-11eb-92e0-c6d8e2b30574.png">

### with gateway

<img width="1423" alt="Screenshot 2021-03-16 at 15 45 28" src="https://user-images.githubusercontent.com/46429701/111320108-55912700-866f-11eb-952a-6fcf95c52f16.png">

**Comment by mcollina** [comment]
I'm not sure the example in your screenshot matches the code. It does not seem to be any input variables in your query.

Can you add a client to your example to reproduce the problem?

**Comment by 0x77dev** [comment]
### Expected behaviour:

`userId: 1` receives: testEvent.userId = 1 

`userId: 2` receives: testEvent.userId = 2

### Client 1 (First Tab)
#### GraphQL Body
```graphql
subscription {
 testEvent {
  userId
 }
}
```
#### connection_init payload
```json
{ "userId": 1 }
```

### Client 2 (Second Tab)
#### GraphQL Body
```graphql
subscription {
 testEvent {
  userId
 }
}
```
#### connection_init payload
```json
{ "userId": 2 }
```

**Comment by mcollina** [comment]
Confirmed. Would you like to send a PR to fix this?

**Comment by 0x77dev** [comment]
I will look at it when I have time

**Comment by simoneb** [comment]
I had a quick look at this and I confirm there is an issue somewhere, even though I haven't been able to nail down were exactly, because the repro does a few things which deviate from a standard scenario.

I forked the original gist here: https://gist.github.com/simoneb/f1a61a1424f8ea7a38cc0606bfb33dcb

This shows that a more typical scenario works correctly:

- remove `onConnect` and `onDisconnect` from the subscription configuration
- remove `resolve` property of the subscription resolver
- use a mutation to trigger the subscription and get the input `userId` from there rather than from HTTP headers

Under these assumptions, the behavior of the gateway is correct by subscribing to a specific user id:

```graphql
subscription {
  testEvent(userId: 2) {
    userId
  }
}
```

and then sending a mutation for the same user id:

```graphql
mutation {
  addTestEvent(userId: 2)
}
```

This is as far as I've got for now. The possible causes are hence either or a combination of:

- `onConnect` subscription configuration
- usage of `resolve` subscription property

Even though it may be a legitimate scenario, I have a feeling that this kind of usage is somewhat unusual though. Before deciding whether to spend any more effort on this, @0x77dev could you please explain what kind of real world scenario this is trying to achieve?

**Comment by 0x77dev** [comment]
@simoneb We are using the `connection_init` payload to authorize requests in our project, in the `connection_init` payload we have token encoded with user id inside, and all our events are scoped to the token.

**Comment by simoneb** [comment]
@0x77dev thanks for the explanation, this clarifies it. I have updated the gist to use `onConnect` to handle that and removed the `userId` parameter from the subscription because the topic the uses subscribes to is inferred from the header as per your original intent. With those changes and my earlier modifications to your example this still works correctly as far as I can see.

Check out the [updated gist](https://gist.github.com/simoneb/f1a61a1424f8ea7a38cc0606bfb33dcb) and use these operations to try it out:

```graphql
// with e.g. { userId: 2 } in headers

subscription {
  testEvent {
    userId
  }
}
```

```graphql
mutation {
  addTestEvent(userId: 2)
}
```

**Comment by simoneb** [comment]
@0x77dev just checking in if you had any news on this. to summarize my last message, a more typical use case which preserves the features of yours seems to work fine.

**Comment by crypto-samba** [comment]
Hi @simoneb ,
Just come across this thread whilst working on a similar issue.  The Gist you've provided doesn't work for me.  If I use graphiql and set headers to {"userId":123}, and add a log line to the subscribe:
`subscribe: async (_, args, { pubsub, userId }) => {
        console.log("subscribing " + userId);
        return await pubsub.subscribe(`testEvent-${userId || 123}`);
      },`

the log shows userId is always undefined.

You also mentioned about how you'd added the onConnect handle.  To advance on the original issue, onConnect is only called once at the start of the app.  This is because (I believe) the gateway multiplexs all the calls down a single websocket, so it's the onConnect is only called when the gateway connects, not when individual clients connect.

Our usecase is similar to above: we have a JWT header for auth/user identification. 

Any help you could offer would be amazing!

**Comment by simoneb** [comment]
@crypto-samba please open a new issue, this has been closed for over a year now
